src/3d/qgstessellator.cpp

```diff
@@ -157,22 +157,87 @@ void QgsTessellator::addPolygon( const QgsPolygonV2 &polygon, float extrusionHei
   polyline.reserve( exterior->numPoints() );
 
   QgsVertexId::VertexType vt;
-  QgsPoint pt, ptPrev;
+  QgsPoint pt;
+
+  QVector3D pNormal( 0, 0, 0 );
+  const int pCount = exterior->numPoints();
+  for ( int i = 0; i < pCount - 1; i++ )
+  {
+    QgsPoint pt1, pt2;
+    exterior->pointAt( i, pt1, vt );
+    exterior->pointAt( i + 1, pt2, vt );
+
+    pNormal.setX( pNormal.x() + ( pt1.y() - pt2.y() ) * ( pt1.z() + pt2.z() ) );
+    pNormal.setY( pNormal.y() + ( pt1.z() - pt2.z() ) * ( pt1.x() + pt2.x() ) );
+    pNormal.setZ( pNormal.z() + ( pt1.x() - pt2.x() ) * ( pt1.y() + pt2.y() ) );
+  }
+
+  pNormal.normalize();
+
+  // Polygon is a triangle
+  if ( pCount == 4 )
+  {
+    QgsPoint pt;
+    for ( int i = 0; i < 3; i++ )
+    {
+      exterior->pointAt( i, pt, vt );
+      mData << pt.x() - mOriginX << pt.z() << - pt.y() + mOriginY;
+      if ( mAddNormals )
+        mData << pNormal.x() << pNormal.z() << - pNormal.y();
+    }
+
+    return;
+  }
+
+  if ( pNormal.length() < 0.999 || pNormal.length() > 1.001 )
+  {
+    return;
+  }
+
+  const QgsPoint ptFirst( exterior->startPoint() );
+  QVector3D pOrigin( ptFirst.x(), ptFirst.y(), ptFirst.z() ), pXVector;
+  // Here we define the two perpendicular vectors that define the local
+  // 2D space on the plane. They will act as axis for which we will
+  // calculate the projection coordinates of a 3D point to the plane.
+  if ( pNormal.z() > 0.001 || pNormal.z() < -0.001 )
+  {
+    pXVector = QVector3D( 1, 0, -pNormal.x() / pNormal.z() );
+  }
+  else if ( pNormal.y() > 0.001 || pNormal.y() < -0.001 )
+  {
+    pXVector = QVector3D( 1, -pNormal.x() / pNormal.y(), 0 );
+  }
+  else
+  {
+    pXVector = QVector3D( -pNormal.y() / pNormal.x(), 1, 0 );
+  }
+  pXVector.normalize();
+  QVector3D pYVector = QVector3D::normal( pNormal, pXVector );
 
-  for ( int i = 0; i < exterior->numPoints() - 1; ++i )
+  for ( int i = 0; i < pCount - 1; ++i )
   {
     exterior->pointAt( i, pt, vt );
-    if ( i == 0 || pt != ptPrev )
+    QVector3D tempPt( pt.x(), pt.y(), ( qIsNaN( pt.z() ) ? 0 : pt.z() ) );
+    const float x = QVector3D::dotProduct( tempPt - pOrigin, pXVector );
+    const float y = QVector3D::dotProduct( tempPt - pOrigin, pYVector );
+
+    const bool found = std::find_if( polyline.begin(), polyline.end(), [x, y]( p2t::Point *&p ) { return *p == *new p2t::Point( x, y ); } ) != polyline.end();
+
+    if ( found )
     {
-      p2t::Point *pt2 = new p2t::Point( pt.x() - mOriginX, pt.y() - mOriginY );
-      polyline.push_back( pt2 );
-      float zPt = qIsNaN( pt.z() ) ? 0 : pt.z();
-      z[pt2] = zPt;
+      continue;
```

Comment:
This is based on the fact that GEOS validation has been skipped (see related comment). Therefore, no check for duplicate points has occurred before and it has to be tested here.

Practically, it can still be a case that a valid GEOS curve could fail this test, in degenerate cases (the projection of non-coplanar points, might end up on duplicate points).

Reply:
@wonder-sk from memory, I don't think that simplicity excludes duplicated points (if one after the other).

Reply:
@vmora you are right - duplicated points one after other are fine for IsSimple() predicate